x/uibc/quota/keeper/quota.go

```diff
@@ -0,0 +1,267 @@
+	params := k.GetParams(ctx)
+
+	quotaOfIBCDenom, err := k.GetQuotaByDenom(ctx, denom)
+	if err != nil {
+		return err
+	}
+
+	if quotaOfIBCDenom == nil {
+		return nil
+	}
+
+	exchangePrice, err := k.getExchangePrice(ctx, denom, amount)
+	if err != nil {
+		// Note: skip the ibc-transfer quota checking if `denom` is not support by leverage
+		if err.Error() == sdkerrors.Wrap(ltypes.ErrNotRegisteredToken, denom).Error() {
+			return nil
+		} else if err != nil {
+			return err
+		}
+	}
+
+	// checking ibc-transfer token quota
+	quotaOfIBCDenom.OutflowSum = quotaOfIBCDenom.OutflowSum.Add(exchangePrice)
+	if quotaOfIBCDenom.OutflowSum.GT(params.TokenQuota) {
+		return uibc.ErrQuotaExceeded
+	}
+
+	// checking total outflow quota
+	totalOutflowSum := k.GetTotalOutflowSum(ctx).Add(exchangePrice)
+	if totalOutflowSum.GT(params.TotalQuota) {
+		return uibc.ErrQuotaExceeded
+	}
+
+	// update the per token outflow sum
+	if err := k.SetDenomQuota(ctx, *quotaOfIBCDenom); err != nil {
+		return err
+	}
+
+	// updating the total outflow sum
+	return k.SetTotalOutflowSum(ctx, totalOutflowSum)
+}
+
+// getExchangePrice
+func (k Keeper) getExchangePrice(ctx sdk.Context, denom string, amount sdkmath.Int) (sdk.Dec, error) {
+	transferCoin := sdk.NewCoin(denom, amount)
+	var err error
+
+	// convert to base asset if it is `uToken`
+	if ltypes.HasUTokenPrefix(denom) {
+		transferCoin, err = k.leverageKeeper.ExchangeUToken(ctx, transferCoin)
+		if err != nil {
+			return sdk.Dec{}, err
+		}
+	}
+
+	// get the exchange price (eg: UMEE) in USD from oracle using base denom eg: `uumee`
+	return k.leverageKeeper.TokenValue(ctx, transferCoin, ltypes.PriceModeSpot)
+}
+
+// UndoUpdateQuota undo the quota of ibc denom
+func (k Keeper) UndoUpdateQuota(ctx sdk.Context, denom string, amount sdkmath.Int) error {
+	quotaOfIBCDenom, err := k.GetQuotaByDenom(ctx, denom)
+	if err != nil {
+		return err
+	}
+
+	if quotaOfIBCDenom == nil {
+		return nil
+	}
+
+	exchangePrice, err := k.getExchangePrice(ctx, denom, amount)
+	if err != nil {
+		// Note: skip the ibc-transfer quota checking if `denom` is not support by leverage
+		if err.Error() == sdkerrors.Wrap(ltypes.ErrNotRegisteredToken, denom).Error() {
+			return nil
+		} else if err != nil {
+			return err
+		}
+	}
+
+	// reset the outflow limit of token
+	quotaOfIBCDenom.OutflowSum = quotaOfIBCDenom.OutflowSum.Sub(exchangePrice)
+	if err := k.SetDenomQuota(ctx, *quotaOfIBCDenom); err != nil {
+		return err
+	}
+
+	// reset the total outflow sum
+	totalOutflowSum := k.GetTotalOutflowSum(ctx)
+	return k.SetTotalOutflowSum(ctx, totalOutflowSum.Sub(exchangePrice))
+}
+
+// GetFundsFromPacket returns transfer amount and denom
+func (k Keeper) GetFundsFromPacket(packet exported.PacketI) (string, string, error) {
+	var packetData transfertypes.FungibleTokenPacketData
+	err := json.Unmarshal(packet.GetData(), &packetData)
+	if err != nil {
+		return "", "", err
+	}
+	return packetData.Amount, k.GetLocalDenom(packetData.Denom), nil
```

Comment:
Can we return `sdk.Int` here? It will help with other functions, and API of this function will be more clear (returning Int rather than string).

Reply:
I will add that